src/DynamicData.Tests/Utilities/TestSourceCache.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
+
+using DynamicData.Kernel;
+
+namespace DynamicData.Tests.Utilities;
+
+public sealed class TestSourceCache<TObject, TKey>
+        : ISourceCache<TObject, TKey>
+    where TObject : notnull
+    where TKey : notnull
+{
+    private readonly IObservable<int> _countChanged;
+    private readonly BehaviorSubject<Exception?> _error;
+    private readonly BehaviorSubject<bool> _hasCompleted;
+    private readonly SourceCache<TObject, TKey> _source;
+
+    public TestSourceCache(Func<TObject, TKey> keySelector)
+    {
+        _error = new(null);
+        _hasCompleted = new(false);
+        _source = new(keySelector);
+
+        _countChanged = WrapStream(_source.CountChanged);
+    }
+
+    public int Count
+        => _source.Count;
+
+    public IObservable<int> CountChanged
+        => _countChanged;
+
+    public IEnumerable<TObject> Items
+        => _source.Items;
+    
+    public IEnumerable<TKey> Keys
+        => _source.Keys;
+
+    public Func<TObject, TKey> KeySelector
+        => _source.KeySelector;
+
+    public IEnumerable<KeyValuePair<TKey, TObject>> KeyValues
+        => KeyValues;
+
+    public void Complete()
+    {
+        AssertCanMutate();
+
+        _hasCompleted.OnNext(true);
+    }
+
+    public IObservable<IChangeSet<TObject, TKey>> Connect(
+            Func<TObject, bool>? predicate = null,
+            bool suppressEmptyChangeSets = true)
+        => WrapStream(_source.Connect(predicate, suppressEmptyChangeSets));
+
+    public void Dispose()
+    {
+        _error.Dispose();
+        _hasCompleted.Dispose();
+        _source.Dispose();
+    }
+
+    public void Edit(Action<ISourceUpdater<TObject, TKey>> updateAction)
+    {
+        AssertCanMutate();
+
+        _source.Edit(updateAction);
+    }
+
+    public Optional<TObject> Lookup(TKey key)
+        => _source.Lookup(key);
+
+    public IObservable<IChangeSet<TObject, TKey>> Preview(Func<TObject, bool>? predicate = null)
+        => WrapStream(_source.Preview(predicate));
+
+    public void SetError(Exception error)
+    {
+        AssertCanMutate();
+
+        _error.OnNext(error);
+    }
+
+    public IObservable<Change<TObject, TKey>> Watch(TKey key)
+        => WrapStream(_source.Watch(key));
+
+    private void AssertCanMutate()
+    {
+        if (_error.Value is not null)
+            throw new InvalidOperationException("The source collection is in an error state and cannot be mutated.");
+
+        if (_hasCompleted.Value)
+            throw new InvalidOperationException("The source collection is in a completed state and cannot be mutated.");
+    }
+
+    private IObservable<T> WrapStream<T>(IObservable<T> sourceStream)
```

Comment:
If you go with the single Subject, then this can just Merge the two together, except Merge won't complete until both child streams complete, so you'd need to have some logic that checks a flag that is set if completion is forced.

Reply:
If swapping to a single subject means I get rid of one clause for handling stream completion, only to replace it with a different one, that doesn't seem like a simplification to me.

Reply:
Okay, I didn't see how many different things were being wrapped with this method.  This is covering ALL the observables, not just Connect.  No, I like this much better now.  It allows you to fail individual subscriptions without having to fail / complete the Subject (which probably isn't recoverable by design).